pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go

```diff
@@ -1841,242 +1841,346 @@ func getNumberOfPodInterfaces(spec *v1.VirtualMachineInstanceSpec) int {
-		errs := validation.IsDNS1123Label(disk.Name)
-
-		for _, err := range errs {
-			causes = append(causes, metav1.StatusCause{
-				Type:    metav1.CauseTypeFieldValueInvalid,
-				Message: err,
-				Field:   field.Child("domain", "devices", "disks").Index(idx).Child("name").String(),
-			})
-		}
+		func(field *k8sfield.Path, idx int, disk v1.Disk) []metav1.StatusCause {
+			return validateDiskNameAsContainerName(field, idx, disk)
+		},
+		// Verify block size is valid
+		func(field *k8sfield.Path, idx int, disk v1.Disk) []metav1.StatusCause {
+			return validateBlockSize(field, idx, disk)
+		},
+	}
 
-		if disk.BlockSize != nil {
-			hasCustomBlockSize := disk.BlockSize.Custom != nil
-			hasVolumeMatchingEnabled := disk.BlockSize.MatchVolume != nil && (disk.BlockSize.MatchVolume.Enabled == nil || *disk.BlockSize.MatchVolume.Enabled)
-			if hasCustomBlockSize && hasVolumeMatchingEnabled {
-				causes = append(causes, metav1.StatusCause{
-					Type:    metav1.CauseTypeFieldValueInvalid,
-					Message: "Block size matching can't be enabled together with a custom value",
-					Field:   field.Index(idx).Child("blockSize").String(),
-				})
-			} else if hasCustomBlockSize {
-				customSize := disk.BlockSize.Custom
-				if customSize.Logical > customSize.Physical {
-					causes = append(causes, metav1.StatusCause{
-						Type:    metav1.CauseTypeFieldValueInvalid,
-						Message: fmt.Sprintf("Logical size %d must be the same or less than the physical size of %d", customSize.Logical, customSize.Physical),
-						Field:   field.Index(idx).Child("blockSize").Child("custom").Child("logical").String(),
-					})
-				} else {
-					checkSize := func(size uint) (bool, string) {
-						if size < 512 {
-							return false, fmt.Sprintf("Provided size of %d is less than the supported minimum size of 512", size)
-						} else if size > 2097152 {
-							return false, fmt.Sprintf("Provided size of %d is greater than the supported maximum size of 2 MiB", size)
-						} else if size&(size-1) != 0 {
-							return false, fmt.Sprintf("Provided size of %d is not a power of 2", size)
-						}
-						return true, ""
-					}
-					if sizeOk, reason := checkSize(customSize.Logical); !sizeOk {
-						causes = append(causes, metav1.StatusCause{
-							Type:    metav1.CauseTypeFieldValueInvalid,
-							Message: reason,
-							Field:   field.Index(idx).Child("blockSize").Child("custom").Child("logical").String(),
-						})
-					}
-					if sizeOk, reason := checkSize(customSize.Physical); !sizeOk {
-						causes = append(causes, metav1.StatusCause{
-							Type:    metav1.CauseTypeFieldValueInvalid,
-							Message: reason,
-							Field:   field.Index(idx).Child("blockSize").Child("custom").Child("physical").String(),
-						})
-					}
-				}
-			}
+	for idx, disk := range disks {
+		for _, validate := range validations {
+			validationCauses := validate(field, idx, disk)
+			causes = append(causes, validationCauses...)
```

Comment:
Since you need to declare the function you can simply append the result of each function to causes without the need to declare an array of functions